### PR TITLE
Support for EC algorithm used in Docker EE client bundles (#887)

### PR DESCRIPTION
**Pull request by chrsoo**
The PR will first try RSA and if this fails fall back to EC which is used in Docker EE client bundles, cf. #887.

Admitedly trying and catching exceptions is not the most elegant way to detect the algorithm to be used for the key spec but there does not seem to be an easy way to dynamically detect the correct algorithm. I guess this is why RSA was hard-coded in the first place...

### CODE REVIEW

**Comment by chrsoo** [comment]
Hi @mattnworb, 

The CONTRIBUTING.md document says to poke you guys if there's no answer so here goes...  would you mind looking at my PR?

File: src/main/java/com/spotify/docker/client/DockerCertificates.java [hunk 1]
```diff
@@ -145,8 +145,26 @@ private PrivateKey readPrivateKey(Path file) throws IOException, InvalidKeySpecE
...
+        kf = KeyFactory.getInstance(algorithm);
+        key = kf.generatePrivate(spec);
+        log.debug("Generated private key from spec using the '{}' algorithm", algorithm);
+        return key;
+      } catch (InvalidKeySpecException | NoSuchAlgorithmException e) {
+        log.debug("Tried generating private key from spec using the '{}' algorithm", algorithm, e);
+      }
+    }
+
+    throw new InvalidKeySpecException("Could not generate private key from spec");
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/DockerCertificates.java]: Can you also include the attempted algorithms in this exception message?

File: src/main/java/com/spotify/docker/client/DockerCertificates.java [hunk 2]
```diff
@@ -145,8 +145,26 @@ private PrivateKey readPrivateKey(Path file) throws IOException, InvalidKeySpecE
   private static PrivateKey generatePrivateKey(PrivateKeyInfo privateKeyInfo) throws IOException,
           InvalidKeySpecException, NoSuchAlgorithmException {
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/DockerCertificates.java]: You can remove `NoSuchAlgorithmException` now.

File: src/main/java/com/spotify/docker/client/DockerCertificates.java [hunk 3]
```diff
@@ -145,8 +145,26 @@ private PrivateKey readPrivateKey(Path file) throws IOException, InvalidKeySpecE
...
     final PKCS8EncodedKeySpec spec = new PKCS8EncodedKeySpec(privateKeyInfo.getEncoded());
-    final KeyFactory kf = KeyFactory.getInstance("RSA");
-    return kf.generatePrivate(spec);
+    return tryGeneratePrivateKey(spec, "RSA", "EC");
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/DockerCertificates.java]: nit: can you put "RSA" and "EC" at the top of the class as a static final ImmutableSet?

like `private static final Set<String> PRIVATE_KEY_ALGS = ImmutableSet.of("RSA", "EC");`

File: src/main/java/com/spotify/docker/client/DockerCertificates.java (hunk 2 above)
Comment (chrsoo) [reply src/main/java/com/spotify/docker/client/DockerCertificates.java]: Done.

File: src/main/java/com/spotify/docker/client/DockerCertificates.java (hunk 3 above)
Comment (chrsoo) [reply src/main/java/com/spotify/docker/client/DockerCertificates.java]: Done.

**Comment by chrsoo** [comment]
Regarding the unit test I added a Docker EE Client Bundle that uses the EC algorithm and copy-pasted the `testReadPrivateKeyPkcs8` test. It passes but I admit to not understanding how it should work. 

As it seems to work even if I change the `key.pem`  key spec that uses the EC algorithm I highly suspect something is off.  Please have a look at the test and help by suggesting any necessary changes!

**Comment by davidxia** [comment]
@chrsoo Thanks for following up. Yea, it looks like the merge has changed some stuff. Could you cherry-pick the relevant commits onto a new branch and PR that's based on the latest upstream master branch? Once you do that it'll be easier for me to review. Thanks again for contributing.

**Pull request closed without merging**